File: python/static/watchlist.js
/* The watchlist: a shortlist that survives the page.
 *
 * Its own file, like kits.js and compare.js - app.js is three and a half
 * thousand lines that cannot be exercised without a whole document around it.
 * See tests/js/test_watchlist.js.
 *
 * Loaded AFTER app.js. Both directions resolve at call time through the global
 * scope classic scripts share:
 *
 *   here needs    currentFplId, esc                  (util.js)
 *                 fixtureTile, shirtImg               (app.js, tables only)
 *   app.js needs  loadWatchlist, toggleWatch, updateWatchButton
 *
 * Stored server-side against the FPL ID, unauthenticated like the drafts - see
 * drafts.py and watchlist.py.
 */

// A Set of codes, so the pop-up's Watch button can answer "is this one on it"
// without a request per player. Refreshed on load and on change, never guessed.
let watchedCodes = new Set();

// The star that appears beside a name in every table and on every profile.
//
// A <button>, not a glyph with a click handler, because it is a control: it
// reaches by keyboard, announces its pressed state, and speaks the same
// vocabulary as #pmWatch in the pop-up. Returned as a string rather than an
// element so it drops into the innerHTML the tables are assembled from -
// handing back a node would mean rewriting all three renderers that use it.
//
// `code` is FPL's season-stable player code. A row that hasn't got one yet -
// every server-rendered row, before /api/all_players lands - gets an empty
// string, so the cell is blank rather than the column going missing and the
// header sliding off the columns underneath it.
function watchStar(code, name) {
    if (code === null || code === undefined) return '';
    const on = watchedCodes.has(code);
    return '<button type="button" class="wl-star' + (on ? ' is-on' : '') + '"'
        + ' data-code="' + esc(code) + '"'
        + ' aria-pressed="' + (on ? 'true' : 'false') + '"'
        + ' title="' + (on ? 'On your watchlist' : 'Add to your watchlist') + '"'
        + ' aria-label="' + (on ? 'Remove ' : 'Add ') + esc(name || 'player')
        + (on ? ' from' : ' to') + ' your watchlist">'
        + (on ? '★' : '☆') + '</button>';
}

// Wire every star inside `root`. Called after each render rather than once at
// load, because these tables rebuild their own rows: a listener bound to a <tr>
// that has since been replaced is bound to nothing.
//
// The click is stopped from propagating deliberately. A star sits inside a row
// whose own click opens the player pop-up, and pressing "watch" should not also
// open a dialog on top of the thing you just pressed.
function bindWatchStars(root) {
    (root || document).querySelectorAll('.wl-star').forEach(btn => {
        if (btn.dataset.wired) return;
        btn.dataset.wired = '1';
        btn.addEventListener('click', ev => {
            ev.stopPropagation();
            ev.preventDefault();
            toggleWatch({ code: +btn.dataset.code }, btn);
        });
    });
}

// One star, repainted in place. The whole table is deliberately not re-rendered
// for a single press: the row the reader just touched would jump if the list
// re-sorted under it, and on the watchlist tab it would vanish mid-click.
function paintStar(btn, on) {
    btn.classList.toggle('is-on', on);
    btn.textContent = on ? '★' : '☆';
    btn.setAttribute('aria-pressed', on ? 'true' : 'false');
    btn.title = on ? 'On your watchlist' : 'Add to your watchlist';
}

function loadWatchlist() {
    const id = currentFplId();
    const body = document.getElementById('watchlistBody');
    if (!body) return Promise.resolve();
    if (!id) {
        body.innerHTML = '<p class="text-muted small mb-0">Enter your FPL ID on the '
            + 'My Team tab first — the watchlist is stored against it.</p>';
        return Promise.resolve();
    }
    body.innerHTML = '<p class="text-muted small mb-0">Loading…</p>';
    return fetch('/api/watchlist/' + id)
        .then(r => r.json())
        .then(d => {
            const players = d.players || [];
            watchedCodes = new Set(players.map(p => p.code));
            renderWatchlist(players);
        })
        .catch(() => {
            body.innerHTML = '<p class="text-muted small mb-0">Couldn’t load the '
                + 'watchlist just now.</p>';
        });
}

function watchlistRow(p) {
    // A departed player keeps his row - dropping it would look like the site
    // had lost the entry. `available` is decided in watchlist.get().
    if (!p.available) {
        return '<tr><td colspan="8" class="text-muted small">'
            + 'Player ' + esc(p.code) + ' is no longer in the game'
            + (p.note ? ' — ' + esc(p.note) : '')
            + ' <button type="button" class="btn btn-link btn-sm p-0 align-baseline wl-remove"'
            + ' data-code="' + esc(p.code) + '">remove</button></td></tr>';
    }
    const gws = (p.next_gameweeks || []).slice(0, 3).map(g => fixtureTile(g)).join('');
    const name = p.path
        ? '<a href="' + esc(p.path) + '">' + esc(p.web_name) + '</a>'
        : esc(p.web_name);
    return '<tr>'
        + '<td class="wl-starcell">' + watchStar(p.code, p.web_name) + '</td>'
        + '<td class="ps-name">' + shirtImg(p.team_code, p.pos, 'shirt-sm')
        + '<span>' + name + '</span></td>'
        + '<td>' + esc(p.pos) + '</td>'
        + '<td class="text-muted">' + esc(p.team_name || '') + '</td>'
        + '<td class="text-end">' + (p.form != null ? esc(p.form) : '–') + '</td>'
        + '<td class="text-end">' + (p.cost != null ? '£' + p.cost.toFixed(1) + 'm' : '') + '</td>'
        + '<td class="wl-drift">' + priceDrift(p) + '</td>'
        + '<td class="text-end">' + (p.predicted != null ? p.predicted.toFixed(1) : '–') + '</td>'
        + '<td class="wl-fixtures">' + gws + '</td>'
        + '<td class="text-end"><button type="button" class="btn btn-link btn-sm p-0 wl-remove"'
        + ' data-code="' + esc(p.code) + '" aria-label="Remove ' + esc(p.web_name) + '">&times;</button></td>'
        + '</tr>';
}

// How close this player is to a price change: a bar and a word for it.
//
// The bar is the share of the threshold his net transfers have covered, and it
// is the same number against the same threshold that the price-changes page
// ranks its risers and fallers on - not a second estimate that could quietly
// disagree with the page that specialises in this.
//
// A player the model cannot measure - one snapshot, or too little ownership for
// the denominator to mean anything - gets a dash. "We cannot say" and "steady"
// are different answers and this column has to be able to give both.
function priceDrift(p) {
    if (!p.price_direction || p.price_progress == null) {
        return '<span class="wl-drift-none" title="Not enough price history for '
            + 'this player yet">–</span>';
    }
    const rising = p.price_direction === 'rise';
    const pct = Math.max(0, Math.min(100, p.price_progress));
    const label = pct >= 100 ? (rising ? 'due a rise' : 'due a fall')
        : pct >= 50 ? (rising ? 'rising' : 'falling')
            : 'steady';
    return '<span class="wl-bar" role="img" aria-label="' + esc(label) + ', '
        + Math.round(pct) + '% of the way to a price change">'
        + '<span class="wl-bar-fill ' + (rising ? 'is-rise' : 'is-fall') + '"'
        + ' style="width:' + pct.toFixed(0) + '%"></span></span>'
        + '<span class="wl-drift-label">' + esc(label) + '</span>';
}

function renderWatchlist(players) {
    const body = document.getElementById('watchlistBody');
    if (!players.length) {
        body.innerHTML = '<p class="text-muted small mb-0">Nothing on your watchlist '
            + 'yet. Open a player from any table and choose <strong>Watch</strong>.</p>';
        return;
    }
    body.innerHTML = '<div class="ps-list"><table class="table table-sm wl-table mb-0">'
        + '<thead><tr><th><span class="visually-hidden">Watching</span></th>'
        + '<th>Player</th><th>Pos</th><th>Club</th>'
        + '<th class="text-end">Form</th><th class="text-end">Price</th>'
        + '<th>Price change</th><th class="text-end">Next GW</th>'
        + '<th>Next 3</th><th></th></tr></thead>'
        + '<tbody>' + players.map(watchlistRow).join('') + '</tbody></table></div>';
    body.querySelectorAll('.wl-remove').forEach(btn =>
        btn.addEventListener('click', () => removeFromWatchlist(+btn.dataset.code)));
    bindWatchStars(body);
}

function removeFromWatchlist(code) {
    const id = currentFplId();
    if (!id) return;
    fetch('/api/watchlist/' + id + '/' + code, { method: 'DELETE' })
        .then(() => { watchedCodes.delete(code); loadWatchlist(); })
        .catch(() => {});
}

function toggleWatch(p, star) {
    const id = currentFplId();
    if (!id) {
        alert('Enter your FPL ID on the My Team tab first — the watchlist is stored against it.');
        return;
    }
    const on = watchedCodes.has(p.code);
    const req = on
        ? fetch('/api/watchlist/' + id + '/' + p.code, { method: 'DELETE' })
        : fetch('/api/watchlist/' + id, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ code: p.code }),
        });
    req.then(r => r.json().catch(() => ({})))
        .then(d => {
            // The cap is enforced server-side; a refusal arrives as `detail`.
            if (d && d.detail) { alert(d.detail); return; }
            if (on) watchedCodes.delete(p.code); else watchedCodes.add(p.code);
            updateWatchButton(p);
            // Every star for this player, wherever it sits on the page. The
            // pop-up usually opens over the very row that produced it, and the
            // two must not disagree once it closes.
            document.querySelectorAll('.wl-star[data-code="' + p.code + '"]')
                .forEach(btn => paintStar(btn, !on));
            if (star) paintStar(star, !on);
            const view = document.getElementById('watchlistView');
            if (view && !view.classList.contains('d-none')) loadWatchlist();
        })
        .catch(() => {});
}

function updateWatchButton(p) {
    const btn = document.getElementById('pmWatch');
    if (!btn) return;
    const on = watchedCodes.has(p.code);
    btn.textContent = on ? '★ Watching' : '☆ Watch';
    btn.classList.toggle('btn-primary', on);
    btn.classList.toggle('btn-outline-primary', !on);
    btn.setAttribute('aria-pressed', on ? 'true' : 'false');
}

// The Watch button must be right the first time a pop-up opens, which is
// usually before the watchlist tab has ever been visited.
let watchedPrimed = false;
function primeWatchedCodes() {
    if (watchedPrimed) return Promise.resolve();
    const id = currentFplId();
    if (!id) return Promise.resolve();
    watchedPrimed = true;
    return fetch('/api/watchlist/' + id)
        .then(r => r.json())
        .then(d => { watchedCodes = new Set((d.players || []).map(p => p.code)); })
        .catch(() => {});
}

File: python/watchlist.py
"""Players a manager is keeping an eye on.

Keyed on `code`, not `element_id`: this is the only thing the app stores that
is meant to outlive a season, and FPL reassigns ids every August.

Unauthenticated in both directions, like the drafts - see the docstring in
drafts.py for that trade. What is stored is a list of publicly listed
footballers plus a note the reader typed, which is why it is acceptable here.
"""

import price_changes
from db import connect, utcnow

# Past a couple of dozen this is a worse copy of the player table - and the
# point at which an unauthenticated write endpoint becomes general storage.
MAX_ENTRIES = 30

# Long enough for "cheap enabler, watch the Newcastle run".
MAX_NOTE_CHARS = 120


class WatchlistError(ValueError):
    """Rejected change - the message is safe to show the user."""


def _clean_note(note):
    """Strip control characters and cap the length.

    Not about injection - the browser escapes on render. About a
    newline-stuffed string making one row a screenful, and about not storing
    unbounded input from an endpoint that takes no credential.
    """
    if note is None:
        return None
    text = "".join(ch if ch.isprintable() else " " for ch in str(note)).strip()
    if not text:
        return None
    if len(text) > MAX_NOTE_CHARS:
        raise WatchlistError(f"Keep the note under {MAX_NOTE_CHARS} characters.")
    return text


def _coerce_code(code):
    try:
        value = int(code)
    except (TypeError, ValueError):
        raise WatchlistError("That is not a player code.")
    if value <= 0:
        raise WatchlistError("That is not a player code.")
    return value


def get(fpl_id, pool=None):
    """The watchlist for `fpl_id`, newest first, joined against the pool.

    Joined at READ time - the row stores only a code, so there is one source of
    truth for who a player is. A code no longer in the pool still comes back
    with `available: False`: dropping it would look like the site had lost the
    entry rather than the player having left.

    Each entry also carries how close the player is to a price change, from the
    same transfer-momentum arithmetic the price-changes page runs - see
    price_changes.for_codes. Asked for the watched codes only, so a thirty-name
    list costs one pass over the snapshot history rather than one per player,
    and absent from the reading means "not measurable", which the table shows
    as a dash rather than as steady.
    """
    by_code = {p["code"]: p for p in (pool or []) if p.get("code") is not None}
    with connect() as conn:
        rows = list(conn.execute(
            """SELECT code, note, added_at FROM watchlist
               WHERE fpl_id = ? ORDER BY added_at DESC, code""", (int(fpl_id),)))

    # Never fatal: a shortlist that cannot say which way a price is drifting is
    # still a shortlist, and the snapshot table is empty for the first two days
    # of any deployment.
    try:
        momentum = price_changes.for_codes([r["code"] for r in rows])
    except Exception as e:
        print(f"couldn't read price momentum for the watchlist: {e}")
        momentum = {}

    out = []
    for r in rows:
        player = by_code.get(int(r["code"]))
        entry = {"code": int(r["code"]), "note": r["note"],
                 "added_at": r["added_at"], "available": player is not None}
        if player:
            entry.update({
                "id": player.get("id"), "web_name": player.get("web_name"),
                "team_name": player.get("team_name"), "team_code": player.get("team_code"),
                "pos": player.get("pos"), "cost": player.get("cost"),
                "predicted": player.get("predicted"), "rating": player.get("rating"),
                "form": player.get("form"), "owned": player.get("owned"),
                "status": player.get("status"), "path": player.get("path"),
                "next_gameweeks": player.get("next_gameweeks") or [],
            })
        drift = momentum.get(int(r["code"]))
        if drift:
            entry.update({
                "price_direction": drift["direction"],
                # Percent of the way to the threshold that has actually moved a
                # price, so the bar means the same thing as the one on the
                # price-changes page.
                "price_progress": drift["progress"],
            })
        out.append(entry)
    return out


def add(fpl_id, code, note=None):
    """Add or update one entry. Idempotent - adding twice updates the note."""
    code = _coerce_code(code)
    note = _clean_note(note)
    with connect() as conn:
        existing = conn.execute(
            "SELECT COUNT(*) FROM watchlist WHERE fpl_id = ?",
            (int(fpl_id),)).fetchone()[0]
        already = conn.execute(
            "SELECT 1 FROM watchlist WHERE fpl_id = ? AND code = ?",
            (int(fpl_id), code)).fetchone() is not None
        # Only for a NEW entry - editing a note must still work at the cap.
        if not already and existing >= MAX_ENTRIES:
            raise WatchlistError(
                f"A watchlist holds {MAX_ENTRIES} players. Remove one first.")
        conn.execute(
            """INSERT INTO watchlist (fpl_id, code, note, added_at)
               VALUES (?, ?, ?, ?)
               ON CONFLICT(fpl_id, code) DO UPDATE SET note = excluded.note""",
            (int(fpl_id), code, note, utcnow()))
    return {"ok": True, "code": code, "note": note}


def remove(fpl_id, code):
    code = _coerce_code(code)
    with connect() as conn:
        removed = conn.execute(
            "DELETE FROM watchlist WHERE fpl_id = ? AND code = ?",
            (int(fpl_id), code)).rowcount
    return {"ok": True, "removed": removed}


def clear(fpl_id):
    with connect() as conn:
        removed = conn.execute(
            "DELETE FROM watchlist WHERE fpl_id = ?", (int(fpl_id),)).rowcount
    return {"ok": True, "removed": removed}
